src/partr.c

```diff
@@ -0,0 +1,1055 @@
+       this task back to the ready queue
+     */
+    if (task->state != done_sym  &&  task->state != failed_sym) {
+        ptls->current_task->next = NULL;
+        JL_LOCK(&task->cq.lock);
+
+        /* ensure the task didn't finish before we got the lock */
+        if (task->state != done_sym  &&  task->state != failed_sym) {
+            /* add the current task to the CQ */
+            if (task->cq.head == NULL) {
+                task->cq.head = ptls->current_task;
+                jl_gc_wb(task, task->cq.head);
+            }
+            else {
+                jl_task_t *pt = task->cq.head;
+                while (pt->next)
+                    pt = pt->next;
+                pt->next = ptls->current_task;
+            }
+
+            JL_UNLOCK(&task->cq.lock);
+
+            jl_task_yield(0);
+        }
+
+        /* the task finished before we could add to its CQ */
+        else
+            JL_UNLOCK(&task->cq.lock);
+    }
+
+    return task->grain_num >= 0 && task->red ?
+           task->red_result : task->result;
+}
+
+
+/*  jl_task_yield() -- cause the invoking task to yield
+
+    If `requeue` is set, the task is inserted into the relevant queue
+    (sticky or multiqueue), otherwise it is assumed it will be re-queued
+    in some other way (e.g. from another task's completion queue).
+ */
+JL_DLLEXPORT void jl_task_yield(int requeue)
+{
+    jl_ptls_t ptls = jl_get_ptls_states();
+    jl_task_t *ytask = ptls->current_task;
+
+#ifdef ENABLE_TIMINGS
+    jl_timing_block_t *blk = ytask->timing_stack;
+    if (blk)
+        jl_timing_block_stop(blk);
+#endif
+
+    if (!jl_setjmp(ytask->ctx, 0)) {
+        ytask->current_tid = -1;
+        ptls->current_task = NULL;
+
+        // backtraces don't survive task switches, see issue #12485
+        ptls->bt_size = 0;
+
+        // save state into yielding task
+        ytask->gcstack = ptls->pgcstack;
+        ytask->world_age = ptls->world_age;
+
+        // If the current task is not holding any locks, free the locks list
+        // so that it can be GC'd without leaking memory.
+        // TODO: this will be too slow!
+        arraylist_t *locks = &ytask->locks;
+        if (locks->len == 0  &&  locks->items != locks->_space) {
+            arraylist_free(locks);
+            arraylist_new(locks, 0);
+        }
+
+        // re-enqueue the task
+        if (requeue)
+            enqueue_task(ytask);
+
+        // run the next available task
+        run_next();
+
+        // unreachable
+        gc_debug_critical_error();
+        abort();
+    }
+
+    // TODO: add support for allowing any thread to run the event loop
+    if (ptls->tid == 0)
+        jl_process_events(jl_global_event_loop());
+
+#ifdef ENABLE_TIMINGS
+    assert(blk == jl_current_task->timing_stack);
+    if (blk)
+        jl_timing_block_start(blk);
+#endif
+}
+
+
+/*  jl_condition_new() -- create a new Condition
+ */
+JL_DLLEXPORT jl_condition_t *jl_condition_new(void)
```

Comment:
I am getting a GC seqfault on a Condition object allocated from here.

1. `JL_GC_PUSH` & `JL_GC_POP` seems to be missing here
2. Is `jl_gc_wb(cond, cond->notify)` necessary? That is a compiler warning for me.

Reply:
Fixed both.